src/ucs/sys/sock.c

```diff
@@ -151,6 +152,76 @@ ucs_status_t ucs_socket_connect_nb_get_status(int fd)
     return UCS_OK;
 }
 
+static inline ucs_status_t ucs_socket_do_io_nb(int fd, void *data, size_t *length_p,
+                                               ucs_socket_io_func_t io_func, const char *name)
+{
+    ssize_t ret;
+
+    ucs_assert(*length_p > 0);
+
+    ret = io_func(fd, data, *length_p, MSG_NOSIGNAL);
+    if (ucs_likely(ret > 0)) {
+        *length_p = ret;
+        return UCS_OK;
+    }
+
+    if (ret == 0) {
+        ucs_trace("fd %d is closed", fd);
+        return UCS_ERR_CANCELED; /* Connection closed */
+    }
+
+    if ((errno == EINTR) || (errno == EAGAIN) || (errno == EWOULDBLOCK)) {
+        return UCS_ERR_NO_PROGRESS;
+    }
+
+    ucs_error("%s(fd=%d data=%p length=%zu) failed: %m",
+              name, fd, data, *length_p);
+    return UCS_ERR_IO_ERROR;
+}
+
+static inline ucs_status_t ucs_socket_do_io_b(int fd, void *data, size_t length,
+                                              ucs_socket_io_func_t io_func, const char *name)
+{
+    size_t done_cnt = 0, cur_cnt = length;
+    ucs_status_t status;
+
+    do {
+        status = ucs_socket_do_io_nb(fd, data, &cur_cnt, io_func, name);
+        if (ucs_unlikely(status != UCS_OK)) {
+            return status;
```

Comment:
Should not we handle `UCS_ERR_NO_PROGRESS` here by `continue` the loop?

Reply:
thanks! fixed